image-rs/src/builder.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2024 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{path::PathBuf, sync::Arc};
+
+use log::info;
+use tokio::sync::RwLock;
+
+use crate::{
+    auth::Auth,
+    config::{ImageConfig, NydusConfig},
+    image::ImageClient,
+    meta_store::{MetaStore, METAFILE},
+    resource::ResourceProvider,
+    signature::ImageSecurity,
+    snapshots::SnapshotType,
+};
+
+use anyhow::Result;
+
+#[derive(Default)]
+pub struct ClientBuilder {
+    config: ImageConfig,
+}
+
+macro_rules! __impl_config {
+    ($name: ident, $type: ident) => {
+        pub fn $name(mut self, $name: $type) -> Self {
+            self.config.$name = $name;
+            self
+        }
+    };
+    ($name: ident, $value: expr, $type: ident) => {
+        pub fn $name(mut self, $name: $type) -> Self {
+            self.config.$name = $value;
+            self
+        }
+    };
+}
+
+impl ClientBuilder {
+    __impl_config!(work_dir, PathBuf);
+    __impl_config!(default_snapshot, SnapshotType);
+    __impl_config!(
+        image_security_policy_uri,
+        Some(image_security_policy_uri),
+        String
+    );
+    __impl_config!(sigstore_config_uri, Some(sigstore_config_uri), String);
+    __impl_config!(
+        authenticated_registry_credentials_uri,
+        Some(authenticated_registry_credentials_uri),
+        String
+    );
+    __impl_config!(max_concurrent_layer_downloads_per_image, usize);
+    __impl_config!(nydus_config, Some(nydus_config), NydusConfig);
+
+    #[cfg(feature = "keywrap-native")]
+    __impl_config!(kbc, String);
+
+    #[cfg(feature = "keywrap-native")]
+    __impl_config!(kbs_uri, kbs_uri, String);
+
+    pub async fn build(self) -> Result<ImageClient> {
+        #[cfg(feature = "keywrap-native")]
+        let resource_provider = Arc::new(ResourceProvider::new(
+            &self.config.kbc,
+            &self.config.kbs_uri,
+            &self.config.work_dir,
+        )?);
+
+        #[cfg(not(feature = "keywrap-native"))]
+        let resource_provider = Arc::new(ResourceProvider::new("", "", &self.config.work_dir)?);
+
+        let registry_auth = match &self.config.authenticated_registry_credentials_uri {
+            Some(uri) => {
+                let auth_bytes = resource_provider.get_resource(uri).await?;
+                let auth = Auth::new(&auth_bytes)?;
+                Some(auth)
+            }
+            None => None,
+        };
+
+        let sigstore_config = match &self.config.sigstore_config_uri {
+            Some(uri) => {
+                let cfg_bytes = resource_provider.get_resource(uri).await?;
+                Some(cfg_bytes)
+            }
+            None => None,
+        };
+
+        let security = match &self.config.image_security_policy_uri {
```

Comment:
also quite a vague name. if we change the name of `ImageSecurity` we can adopt something similar here.

Reply:
How about `image_security_module`?